torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -398,19 +411,63 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
       : bdimx + (device_warp_size - bdimx % device_warp_size);
 
   bool pad_bdimx = bdimx > 16 &&
-      padded_bdimx * bdimy * bdimz <
-          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
+      padded_bdimx * bdimy * bdimz < (int64_t)dev_prop->maxThreadsPerBlock;
 
   pad_bdimx = pad_bdimx &&
       bdimx * inner_reduction_unroll_factor != inner_most_dimension_numel;
 
+  int64_t nvrtc_register_per_thread = LaunchParams::UNINITIALIZED_VAL;
+  const int blocksPerKernel = godim;
+  if (vectorize && blocksPerKernel > device_multiprocessor_count &&
+      batches_per_block_inner_reduction > 1) {
+    // estimate register usage and occupancy raito.
+    // If occupancy raito is less than occupancy_ratio, reduce register usage
+    constexpr int reg_allocation_granularity = 256;
+    constexpr float occupancy_ratio = 0.4;
+    const int persistent_buffer_size = batches_per_block_inner_reduction *
+        inner_reduction_unroll_factor * max_input_dtype_size;
+    // persistent_buffer_size = 4*2, 8*2, 32*2, 64*2, 128*2
+    // register_used_on_a100  = 27,  40,  62,   73,   105
+    // register_used_on_v100  = xx,  xx,  45,   62,   93
+    // estimated_register_num = 42,  44,  56,   72,   104
+    // safe for both v100 & a100
+    const int estimated_register_count = persistent_buffer_size / 4 + 40;
```

Comment:
Is the division by 4 because of 8-way vectorization and half data type? If so, it should be parameterized by those factors.

Reply:
constexpr int bytes_per_register = 4;